### DISCUSSION TITLE
Fastapi Dependencies (and integration with APScheduler)

### DISCUSSION THREAD

**Discussion by j-4ck** [fastapi-users/fastapi-users#1471] [Q&A] [answered]
Hey all, I'm looking for some advice with a problem I have encountered over the last few days. The situation I am currently in (I will attach code snippets below) is I have an async sqlalchemy DB session, which I access using FastAPI's Depends in my frontend (NiceGUI) and backend. I need to be able to access DB sessions not only from a non-FastAPI function, but from an APScheduler background function. I haven't yet had any success, and I was wondering am I approaching this incorrectly? I can yield DB sessions from a FastAPI function/endpoint (such as /getSession) and pass it down the function tree, but then I encounter errors with APScheduler since I am trying to call async code in a sync function. My other idea was to create an endpoint that will do the DB leg-work for me (such as /setValues), but again calling this from my background schedule results in runtime errors. Any advice would be welcome.
Below is my simplified process of yielding a database session:
![image](https://github.com/user-attachments/assets/41b20c61-96ab-4758-8a2f-4373722fe7c8)
Below is my simplified background process:
![image](https://github.com/user-attachments/assets/ac9eebb9-f212-4032-988c-aac902ac81a8)

I am new to FastAPI, so I am open to redesigning the flow if I am approaching this incorrectly. Let me know if you need me to share any more information.

Edit: Ignore the passing of value "db" to job() in add_job. Also wanted to clarify not necessarily looking for a direct solution here, more of an open discussion on the "proper" method? Thanks

**Comment by hgalytoby** [comment]
Have you tried the `asyncio.run()` method?

**Reply by j-4ck** [reply, marked as answer]
I did, but without success. I'm now closing the discussion as I have found a working method using APSchedulers AsyncIOScheduler. This allowed me to then pass my db session down the function tree and successfully await the necessary db functions in my job(). A simple solution that I probably should've realised from the beginning
